gravitee-am-gateway/gravitee-am-gateway-handler/gravitee-am-gateway-handler-core/src/main/java/io/gravitee/am/gateway/handler/root/resources/handler/login/LoginSelectionRuleHandler.java

```diff
@@ -0,0 +1,110 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.gravitee.am.gateway.handler.root.resources.handler.login;
+
+import com.google.common.base.Strings;
+import io.gravitee.am.common.oidc.Parameters;
+import io.gravitee.am.common.utils.ConstantKeys;
+import io.gravitee.am.common.web.UriBuilder;
+import io.gravitee.am.gateway.handler.common.vertx.core.http.VertxHttpServerRequest;
+import io.gravitee.am.identityprovider.api.SimpleAuthenticationContext;
+import io.gravitee.am.model.IdentityProvider;
+import io.gravitee.am.model.idp.ApplicationIdentityProvider;
+import io.gravitee.am.model.oidc.Client;
+import io.vertx.core.Handler;
+import io.vertx.reactivex.ext.web.RoutingContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static io.gravitee.am.common.utils.ConstantKeys.USERNAME_PARAM_KEY;
+import static io.gravitee.am.gateway.handler.root.resources.handler.login.LoginSocialAuthenticationHandler.SOCIAL_AUTHORIZE_URL_CONTEXT_KEY;
+import static io.gravitee.am.gateway.handler.root.resources.handler.login.LoginSocialAuthenticationHandler.SOCIAL_PROVIDER_CONTEXT_KEY;
+
+/**
+ * @author Eric LELEU (eric.leleu at graviteesource.com)
+ * @author GraviteeSource Team
+ */
+public class LoginSelectionRuleHandler  implements Handler<RoutingContext>  {
+    private final Logger logger = LoggerFactory.getLogger(this.getClass());
+
+    private final boolean fromIdentifierFirstLogin;
+
+    public LoginSelectionRuleHandler(boolean fromIdentifierFirstLogin) {
+        this.fromIdentifierFirstLogin = fromIdentifierFirstLogin;
+    }
+
+    @Override
+    public void handle(RoutingContext routingContext) {
+        final Client client = routingContext.get(ConstantKeys.CLIENT_CONTEXT_KEY);
+
+        List<IdentityProvider> socialProviders = routingContext.get(SOCIAL_PROVIDER_CONTEXT_KEY);
+        if ((socialProviders != null && !socialProviders.isEmpty())
+                && (client.getIdentityProviders() != null && !client.getIdentityProviders().isEmpty())) {
+
+            var appIdpMap = client.getIdentityProviders().stream().collect(Collectors.toMap(
+                    ApplicationIdentityProvider::getIdentity, Function.identity()
+            ));
+
+            var context = new SimpleAuthenticationContext(new VertxHttpServerRequest(routingContext.request().getDelegate()), routingContext.data());
+            var templateEngine = context.getTemplateEngine();
+            var identityProvider = socialProviders.stream()
+                    .filter(idp -> appIdpMap.containsKey(idp.getId()))
+                    .filter(idp -> evaluateRule(appIdpMap.get(idp.getId()), templateEngine, idp))
+                    .findFirst();
+
```

Comment:
In order to keep the order of the ordered set, we need to switch the social provider collection and the map

```suggestion
        var socialProviderMap = socialProviders.stream().collect(Collectors.toMap(
                    IdentityProvider::getId, Function.identity()
            ));

            var context = new SimpleAuthenticationContext(new VertxHttpServerRequest(routingContext.request().getDelegate()), routingContext.data());
            var templateEngine = context.getTemplateEngine();
            var identityProvider = client.getIdentityProviders().stream()
                    .filter(appIdp -> socialProviderMap.containsKey(appIdp.getIdentity()))
                    .filter(appIdp -> evaluateRule(appIdp, templateEngine))
                    .findFirst();

```